articles/055_ros_parameter_design.md

```diff
@@ -0,0 +1,184 @@
+
+## Ideal System
+
+It is useful to consider the ideal system to understand how it relates to the current system and how a new system could work.
+We'd like to support both the use cases of standard parameters as well as dynamic parameters.
+An ideal parameter system would have the qualities laid out in the following paragraphs.
+
+### Accept parameter values
+
+At the most basic a parameter system must be able to accept the setting of parameter values.
+An extension is to accept a group of parameters updated atomically.
+This allows you to update coupled parameters such as PID gains without worring about invalid intermediate states.
+
+### Return parameter values
+
+Also at the core functionality return the value of one or more parameters which have previously been set.
+
+### Unset a parameter value
+
+If a parameter has been set it should be able to be unset.
+
+### Introspection of available parameters
+
+Provide a list of currently set parameters.
+
+### Notifications of change
+
+When a parameter changes value it should be possible to get a notification.
+
+### Reject parameter changes
+
+If a parameter change is requested to an invalid value the parameter server should be able to reject the change.
+Related to this it should also be able to communicate the preconditions which are necessary but not sufficient to determine the acceptance of a change.
+
+### Introspection of expected parameters
+
+When using parameters providing a list of expected parameters can prevent accidentally setting parameters with typos in their names etc.
+
+### Control parameter lifetime
+
+When parameters are set, there should be an understanding of what the lifetime of the parameter value will be and what conditions will clear its value.
+
+### Unambigious naming
+
+All parameters should be unambiguously addressable.
+
+For example one of the challenges of the current system is that there is a naming ambiguity between nodes and parameters `/foo/bar/baz` could be a node `/foo/bar/baz` or a private parameter `baz` on node `/foo/bar`.
+
+### Logging
+
+All changes to parameters should be visible when logging.
+And when playing back a log file the parameter changes should be able to be reapplied.
+
+## Proposed Approach
+
+To cover the feature set above the ROS 2.0 parameter system is proposed as follows.
+
+### Parameters Hosted on nodes
+
+For clarity of parameter lifecycle and validation all parameters will be hosted on a node.
+Their lifetime will be implicitly tied to the nodes lifetime.
+The node will be responsible for validating current values.
+
+### Parameter addressing
+
+All parameters will be addressed by two elements, the full node name, and the parameter name.
+
+### Supported datatypes
+
+Each parameter consists of a key and a value.
+The key is a string value.
+The value can be one of the following datatypes:
+
+- `float64`
+- `int64`
+- `string`
+- `bool`
+- `bytes[]`
```

Comment:
In general shorter bit values are important for space and speed optimization. However supporting those optimizations increases the complexity of the interface. In this case it would close to double the number of datatypes supported. People would then have to be careful about up and down conversions and range checks which becomes noteably harder when interacting with loosely typed languages.